src/Core/Resolvers/MsSqlQueryBuilder.cs

```diff
@@ -63,27 +65,170 @@ public string Build(SqlQueryStructure structure)
+        /// </summary>
+        /// <param name="sourceDefinition">Table definition.</param>
+        private static (string, List<string>) GetSegregatedPKColumns(SourceDefinition sourceDefinition)
+        {
+            string autoGenPKColumn = string.Empty;
+            List<string> nonAutoGenPKColumns = new();
+            foreach (string primaryKey in sourceDefinition.PrimaryKey)
+            {
+                if (sourceDefinition.Columns[primaryKey].IsAutoGenerated)
+                {
+                    autoGenPKColumn = primaryKey;
+                }
+                else
+                {
+                    nonAutoGenPKColumns.Add(primaryKey);
+                }
+            }
+
+            return (autoGenPKColumn, nonAutoGenPKColumns);
         }
 
         /// <inheritdoc />
         public string Build(SqlUpdateStructure structure)
         {
+            SourceDefinition sourceDefinition = structure.GetUnderlyingSourceDefinition();
+            bool isUpdateTriggerEnabled = sourceDefinition.IsUpdateDMLTriggerEnabled;
+            string tableName = $"{QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)}";
             string predicates = JoinPredicateStrings(
                                    structure.GetDbPolicyForOperation(EntityActionOperation.Update),
                                    Build(structure.Predicates));
+            string columnsToBeReturned =
+                MakeOutputColumns(structure.OutputColumns, isUpdateTriggerEnabled ? string.Empty : OutputQualifier.Inserted.ToString());
+
+            StringBuilder updateQuery = new($"UPDATE {tableName} SET {Build(structure.UpdateOperations, ", ")} ");
 
-            return $"UPDATE {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} " +
-                    $"SET {Build(structure.UpdateOperations, ", ")} " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"WHERE {predicates};";
+            // If a trigger is enabled on the entity, we cannot use OUTPUT clause to return the record.
+            // In such a case, we will use a subsequent select query to get the record. By the time the subsequent select executes,
+            // the trigger would have already executed and we get the data as it is present in the table.
+            if (isUpdateTriggerEnabled)
+            {
+                updateQuery.Append($"WHERE {predicates};");
+                updateQuery.Append($"SELECT {columnsToBeReturned} FROM {tableName} WHERE {predicates};");
+                /* An example final update query on a table with update trigger enabled would look like:
+                 *  UPDATE [dbo].[table] SET [dbo].[table].[field1] = @param3
+                 *  WHERE ([field3] != @param0) AND [dbo].[table].[id] = @param1 AND [dbo].[table].[nonautogen_id] = @param2;
```

Comment:
`([field3] != @param0)` is an example of a predicate representing the DB Policy?

Reply:
Right!